From 54e9087da58e7ce048c32ab1013403f1cdf2d297 Mon Sep 17 00:00:00 2001
From: Myron Stowe <myron.stowe@redhat.com>
Date: Wed, 6 Aug 2014 03:48:17 -0400
Subject: [x86] pci: Set IORESOURCE_ROM_SHADOW only for the default VGA device

Message-id: <20140806034817.1438.27308.stgit@gir.stowe>
Patchwork-id: 87448
O-Subject: [RHEL7.1 PATCH 39/69] x86/PCI: Set IORESOURCE_ROM_SHADOW only for the default VGA device
Bugzilla: 1110896
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Dean Nelson <dnelson@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1110896
Upstream Status: d8801e4df91ee788486a11b36f45aa04784f7b4a

commit d8801e4df91ee788486a11b36f45aa04784f7b4a
Author: Sander Eikelenboom <linux@eikelenboom.it>
Date:   Fri Jan 31 10:28:23 2014 +0100

    x86/PCI: Set IORESOURCE_ROM_SHADOW only for the default VGA device

    Setting the IORESOURCE_ROM_SHADOW flag on a VGA card other than the primary
    prevents it from reading its own ROM.  It will get the content of the
    shadow ROM at C000 instead, which is of the primary VGA card and the driver
    of the secondary card will bail out.

    Fix this by checking if the arch code or vga-arbitration has already
    determined the vga_default_device, if so only apply the fix to this primary
    video device and let the comment reflect this.

    [bhelgaas: add subject, split x86 & ia64 into separate patches]
    Signed-off-by: Sander Eikelenboom <linux@eikelenboom.it>
    Signed-off-by: Bjorn Helgaas <bhelgaas@google.com>
    Reviewed-by: Konrad Rzeszutek Wilk <konrad.wilk@oracle.com>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/x86/pci/fixup.c b/arch/x86/pci/fixup.c
index 201833f..94ae9ae 100644
--- a/arch/x86/pci/fixup.c
+++ b/arch/x86/pci/fixup.c
@@ -313,9 +313,10 @@ DECLARE_PCI_FIXUP_FINAL(PCI_VENDOR_ID_INTEL, PCI_DEVICE_ID_INTEL_MCH_PC1, pcie_r
  * IORESOURCE_ROM_SHADOW is used to associate the boot video
  * card with this copy. On laptops this copy has to be used since
  * the main ROM may be compressed or combined with another image.
- * See pci_map_rom() for use of this flag. IORESOURCE_ROM_SHADOW
- * is marked here since the boot video device will be the only enabled
- * video device at this point.
+ * See pci_map_rom() for use of this flag. Before marking the device
+ * with IORESOURCE_ROM_SHADOW check if a vga_default_device is already set
+ * by either arch cde or vga-arbitration, if so only apply the fixup to this
+ * already determined primary video card.
  */
 
 static void pci_fixup_video(struct pci_dev *pdev)
@@ -346,12 +347,13 @@ static void pci_fixup_video(struct pci_dev *pdev)
   }
   bus = bus->parent;
  }
- pci_read_config_word(pdev, PCI_COMMAND, &config);
- if (config & (PCI_COMMAND_IO | PCI_COMMAND_MEMORY)) {
-  pdev->resource[PCI_ROM_RESOURCE].flags |= IORESOURCE_ROM_SHADOW;
-  dev_printk(KERN_DEBUG, &pdev->dev, "Boot video device\n");
-  if (!vga_default_device())
+ if (!vga_default_device() || pdev == vga_default_device()) {
+  pci_read_config_word(pdev, PCI_COMMAND, &config);
+  if (config & (PCI_COMMAND_IO | PCI_COMMAND_MEMORY)) {
+   pdev->resource[PCI_ROM_RESOURCE].flags |= IORESOURCE_ROM_SHADOW;
+   dev_printk(KERN_DEBUG, &pdev->dev, "Boot video device\n");
    vga_set_default_device(pdev);
+  }
  }
 }
 DECLARE_PCI_FIXUP_CLASS_FINAL(PCI_ANY_ID, PCI_ANY_ID,
-- 
1.7.1